frontend/pkg/frontend/frontend.go

```diff
@@ -403,6 +386,114 @@ func (f *Frontend) ArmResourceList(writer http.ResponseWriter, request *http.Req
 	}
 }
 
+func cosmosListOptions(resourceType *azcorearm.ResourceType, queryParams url.Values) database.DBClientListResourceDocsOptions {
+	options := database.DBClientListResourceDocsOptions{
+		ResourceType: resourceType,
+		PageSizeHint: api.Ptr(int32(20)),
+	}
+	// The Resource Provider Contract implies $top is only honored when
+	// following a "nextLink" after the initial collection GET request.
+	// So only check for it when the URL includes a $skipToken.
+	if queryParams.Has("$skipToken") {
+		options.ContinuationToken = api.Ptr(queryParams.Get("$skipToken"))
+		top, err := strconv.ParseInt(queryParams.Get("$top"), 10, 32)
+		if err == nil && top > 0 {
+			options.PageSizeHint = api.Ptr(int32(top))
+		}
+	}
+	return options
+}
+
+func (f *Frontend) ArmResourceListHCPClusters(writer http.ResponseWriter, request *http.Request) {
```

Comment:
This is the worst one.  Once cluster-service is gone, a per-item transform would make it about 50% tighter.